🧰 Rewrite _mint in Yul

0xkarmacoma opened the pull request:
The mint 1000 benchmark now takes -25408 gas (-0.4%)

Savings come from:
- avoiding a spurious SLOAD when we store the primary pointer
- cleaning the upper bits of the pointer address only when necessary
- computing the data offset of `addresses` outside of the loop
- performing a single extcodesize check on the pointer
- inlining the SSTORE2.read operation

0xkarmacoma committed:
Save the data pointer out of the loop in mint

plus use more descriptive names to make the assembly more readable

test_mint_newcomer_1000() (gas: -5986 (-0.095%))

0xkarmacoma committed:
Add assembly comments

0xkarmacoma committed:
Avoid an SLOAD when setting the primary owner pointer

Saves 145 gas

0xkarmacoma committed:
Keep _safeMint in sync with _mint

0xkarmacoma committed:
Go full Yul in _mint, full Solidity in _safeMint

-374 gas for _mint
+191 gas for _safeMint

0xkarmacoma committed:
Remove redundant check

jorgeavaldez commented on src/SS2ERC721.sol:
```diff
@@ -119,7 +129,7 @@ abstract contract SS2ERC721 is ERC721 {
         balance = balanceIndicator & BALANCE_MASK;
 
-        if (balanceIndicator & NOT_PRIMARY_OWNER_FLAG == 0) {
+        if (balanceIndicator & SKIP_PRIMARY_BALANCE == 0) {
```
`SKIP_PRIMARY_BALANCE` is much more intuitive in this context

jorgeavaldez commented on src/SS2ERC721.sol:
```diff
@@ -183,30 +193,82 @@ abstract contract SS2ERC721 is ERC721 {
+            let pointer_codesize := extcodesize(clean_pointer)
+            if lt(pointer_codesize, 2) {
+                revert_invalid_addresses()
+            }
```
As discussed in slack:
- if the address has no code, then it's not an SSTORE2 pointer
- if the address has 1 byte of code, then it could be a valid but empty SSTORE2 pointer

1 byte because SSTORE2 prepends the buffer with 00, which corresponds to the STOP opcode

0xkarmacoma replied:
added in 9f1783a 👍

jorgeavaldez commented on src/SS2ERC721.sol:
```diff
@@ -183,30 +193,82 @@ abstract contract SS2ERC721 is ERC721 {
+                addresses_length        // size
+            )
+
+            numMinted := div(addresses_length, ADDRESS_SIZE_BYTES)
```
I know this isn't the hotpath but just wondering if it's possible.

We perform a division both here and on 239. Is it possible to divide and set a remainder flag?

jorgeavaldez replied:
Didn't see it here so i assume no: https://ethereum.github.io/execution-specs/autoapi/ethereum/frontier/vm/instructions/arithmetic/index.html#div

0xkarmacoma replied:
that would be neat but I don't think there is direct support for that in the EVM. There may be a clever bit twiddling solution, but as you pointed out it wouldn't save that much

jorgeavaldez commented on src/SS2ERC721.sol:
```diff
@@ -183,30 +193,82 @@ abstract contract SS2ERC721 is ERC721 {
+                let to_ptr := add(addresses_data, mul(i, ADDRESS_SIZE_BYTES))
+
+                // load and shift the address to the right by 96 bits:
+                //      before: ADDR_ADDR_ADDR_ADDR_ADDR_ADDR_ADDR_ADDR_XXXXXXXXXXXXXXXXXXXXXXXX
+                //      after:  000000000000000000000000_ADDR_ADDR_ADDR_ADDR_ADDR_ADDR_ADDR_ADDR
```
Curious if we can pack them in this way ahead of time to avoid this operation

jorgeavaldez replied:
I see this is because mload doesn't let you control size. Could we use extcodecopy instead?

0xkarmacoma replied:
in principle yes, but extcodecopy is very expensive (100 gas minimum) so we frontload it and then process everything in memory with cheap mloads and shifts (3 gas each)

jorgeavaldez reviewed (approved):
Logically, it checks out! No red flags! Just some questions, maybe we can optimize a lil more? :stuck_out_tongue_closed_eyes: 

LGTM :fire: :shipit:

0xkarmacoma committed:
Add more comments in the assembly

0xkarmacoma committed:
Fix off by one tokenId error

i must be incremented after the pointer arithmetic but before the event emission
